--- a/tools/winebuild/utils.c	2023-07-19 18:29:22.000000000 +0300
+++ b/tools/winebuild/utils.c	2026-06-01 17:28:57.213784494 +0300
@@ -381,6 +381,10 @@
             strarray_add( &args, "-m" );
             strarray_add( &args, (force_pointer_size == 8) ? "i386pep" : "i386pe" );
             break;
+        case PLATFORM_SOLARIS:
+            strarray_add( &args, "-m" );
+            strarray_add( &args, (force_pointer_size == 8) ? "elf_x86_64_sol2" : "elf_i386_sol2" );
+            break;
         default:
             strarray_add( &args, "-m" );
             strarray_add( &args, (force_pointer_size == 8) ? "elf_x86_64" : "elf_i386" );
