--- a/configure	2023-07-19 18:29:22.000000000 +0300
+++ b/configure	2026-06-01 18:34:39.403698138 +0300
@@ -10093,6 +10093,13 @@
     fi
     ;;
 
+  solaris2*)
+    DLLFLAGS="$DLLFLAGS -fPIC"
+    LDDLLFLAGS="-fPIC"
+    LIBWINE_LDFLAGS="-shared -Wl,-soname,libwine.so.$libwine_soversion"
+    WINEPRELOADER_LDFLAGS="-static -nostartfiles -nodefaultlibs -Wl,-Ttext=0x7d400000"
+    ;;
+
   *)
 
 printf "%s\n" "#define _GNU_SOURCE 1" >>confdefs.h
@@ -10102,6 +10109,9 @@
 
     case $host_cpu in
       *i[3456789]86*)
+        case $host_os in
+          solaris2*) ;;
+          *)
         DLLFLAGS="$DLLFLAGS -fno-PIC"
         LDDLLFLAGS="-fno-PIC"
         { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking whether the compiler supports -fno-PIC -Wl,-z,notext" >&5
@@ -10133,6 +10143,8 @@
   LDDLLFLAGS="$LDDLLFLAGS -Wl,-z,notext"
 fi
         ;;
+        esac
+        ;;
       *)
         DLLFLAGS="$DLLFLAGS -fPIC"
         LDDLLFLAGS="-fPIC" ;;
@@ -21531,6 +21543,10 @@
 wine_fn_config_makefile dlls/npptools enable_npptools
 wine_fn_config_makefile dlls/nsi enable_nsi
 wine_fn_config_makefile dlls/nsi/tests enable_tests
+# OpenIndiana/illumos: nsiproxy.sys needs Linux-specific network ioctls
+case $host_os in
+solaris2*) enable_nsiproxy_sys=no ;;
+esac
 wine_fn_config_makefile dlls/nsiproxy.sys enable_nsiproxy_sys
 wine_fn_config_makefile dlls/ntdll enable_ntdll
 wine_fn_config_makefile dlls/ntdll/tests enable_tests
